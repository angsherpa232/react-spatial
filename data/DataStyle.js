import Style from 'ol/style/Style';
import Circle from 'ol/style/Circle';
import Stroke from 'ol/style/Stroke';
import Fill from 'ol/style/Fill';

const data = {
  child31: new Style({
    image: new Circle({
      radius: 5,
      fill: new Fill({
        color: '#ff0000',
      }),
    }),
  }),
  child32: new Style({
    stroke: new Stroke({
      color: '#ffcc33',
      width: 2,
    }),
  }),
  child33: new Style({
    stroke: new Stroke({
      color: '#7dff8f',
      width: 3,
    }),
    fill: new Fill({
      color: 'rgba(125, 255, 143, 0.2)',
    }),
  }),
};

export default data;
