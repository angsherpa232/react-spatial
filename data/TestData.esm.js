import { applyDefaultValues } from './TreeData.esm';

const testData = {
  rootId: 'root',
  items: {
    root: {
      id: 'root',
      children: ['topic1'],
      type: 'checkbox',
      data: {
        title: 'root',
      },
    },
    topic1: {
      id: 'topic1',
      isChecked: true,
      isExpanded: true,
      type: 'radio',
      children: ['baselayer1', 'sublayers1', 'sublayers2'],
      data: {
        title: 'Topic 1',
      },
    },
    baselayer1: {
      id: 'baselayer1',
      type: 'checkbox',
      isChecked: true,
      data: {
        title: 'OSM Baselayer',
        type: 'xyz',
      },
    },
    sublayers1: {
      id: 'sublayers1',
      isChecked: true,
      isExpanded: true,
      type: 'radio',
      children: ['usaPop', 'countryBorders'],
      data: {
        title: 'Sub-Layers 1',
      },
    },
    usaPop: {
      id: 'usaPop',
      type: 'checkbox',
      isChecked: true,
      defaults: {
        isChecked: true,
      },
      data: {
        title: 'USA Population Density',
        type: 'wmts',
      },
    },
    countryBorders: {
      id: 'countryBorders',
      type: 'checkbox',
      isChecked: false,
      data: {
        title: 'Countries Borders',
        type: 'vectorLayer',
      },
    },
    sublayers2: {
      id: 'sublayers2',
      isChecked: false,
      isExpanded: true,
      type: 'radio',
      children: ['pointLayer', 'lineLayer', 'polygonLayer'],
      data: {
        title: 'Sub-Layers 2',
      },
    },
    pointLayer: {
      id: 'pointLayer',
      type: 'radio',
      isChecked: false,
      defaults: {
        isChecked: true,
      },
      data: {
        title: 'Points Samples',
        type: 'vectorLayer',
      },
    },
    lineLayer: {
      id: 'lineLayer',
      type: 'radio',
      isChecked: false,
      data: {
        title: 'Lines Samples',
        styleId: 'lineLayer',
        type: 'vectorLayer',
      },
    },
    polygonLayer: {
      id: 'polygonLayer',
      type: 'radio',
      isChecked: false,
      data: {
        title: 'Polygons Samples',
        type: 'vectorLayer',
      },
    },
  },
};

export default applyDefaultValues(testData);
