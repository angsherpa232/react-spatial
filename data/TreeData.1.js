/**
 *  Node definition:
 *
 *     id: 'root',
 *     children: ['ovnetzkarteschweiz', 'bauprojekte'],
 *     isExpanded: true,
 *     isChecked: false,
 *     isChildrenLoading: false,
 *     type: 'checkbox',
 *     default: {
 *        isExpanded: true,
 *        isChecked: false
 *     },
 *     data: {
 *       title: 'root',
 *     },
 *
 *     // experimental
 *     useDefaults: true,
 *     default: {
 *        isExpanded: true,
 *        isChecked: false
 *     },
 *
 */

const data = {
  rootId: 'root',
  items: {
    root: {
      id: 'root',
      children: ['ovnetzkarteschweiz', 'bauprojekte'],
      type: 'checkbox',
      data: {
        title: 'root',
      },
    },
    ovnetzkarteschweiz: {
      id: 'ovnetzkarteschweiz',
      isExpanded: true,
      children: ['bahnhofplane', 'passagierfrequenzen'],
      type: 'radio',
      data: {
        title: 'ovnetzkarteschweiz',
      },
    },
    bauprojekte: {
      id: 'bauprojekte',
      type: 'radio',
      data: {
        title: 'bauprojekte',
      },
    },
    bahnhofplane: {
      id: 'bahnhofplane',
      isExpanded: true,
      children: ['interaktivbahnhoplan', 'printprodukte'],
      type: 'checkbox',
      data: {
        title: 'bahnhofplane',
      },
    },
    passagierfrequenzen: {
      id: 'passagierfrequenzen',
      children: ['interaktivbahnhoplan2', 'printprodukte2'],
      type: 'checkbox',
      data: {
        title: 'passagierfrequenzen',
      },
    },
    interaktivbahnhoplan: {
      id: 'interaktivbahnhoplan',
      type: 'radio',
      data: {
        title: 'interaktivbahnhoplan',
      },
    },
    printprodukte: {
      id: 'printprodukte',
      type: 'radio',
      data: {
        title: 'printprodukte',
      },
    },
    interaktivbahnhoplan2: {
      children: ['interaktivbahnhoplan3', 'printprodukte3'],
    },
    printprodukte2: {},
    interaktivbahnhoplan3: {},
    printprodukte3: {},
  },
};

const applyDefaultValuesOnItem = (d, id) => {
  const item = d.items[id];
  if (!item) {
    // eslint-disable-next-line no-console
    console.error(`No item with id  ${id}`);
    return;
  }
  item.id = item.id || id;
  item.type = item.type || 'checkbox';
  item.data = item.data || {};
  item.data.title = item.data.title || id;
  item.hasChildren = !!(item.children && item.children.length);
  item.hasParent = item.hasParent || false;
  item.parentId = item.parentId || null;
  item.isExpanded = item.isExpanded || false;
  item.isChecked = item.isChecked || false;
  item.isChildrenLoading = item.isChildrenLoading || false;
};

// Fill the data tree with some helpers properties
const applyDefaultValues = dat => {
  const d = { ...dat };
  Object.keys(d.items).forEach(id => {
    applyDefaultValuesOnItem(d, id);
    const item = d.items[id];
    if (item.hasChildren) {
      item.hasChildren = true;
      item.children.forEach(childId => {
        applyDefaultValuesOnItem(d, childId);
        d.items[childId].parentId = id;
        d.items[childId].hasParent = true;
      });
    } else {
      item.children = [];
    }
  });
  return d;
};

// For styleguidist, see styleguide.config.js
/* if (module.exports) {
  module.exports = {
    data: applyDefaultValues(data),
    applyDefaultValues,
  };
} */

// For tests
export { applyDefaultValues };
export default applyDefaultValues(data);
