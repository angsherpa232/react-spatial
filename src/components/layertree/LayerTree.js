import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Tree, { mutateTree } from '@atlaskit/tree';

const propTypes = {
  /**
   * The object representing a hierachical object to display as a tree.
   * This property is only used at initialisation time. Use onItemChange if
   * you want to keep track of the tree status.
   */
  tree: PropTypes.shape({
    rootId: PropTypes.string.isRequired,
    items: PropTypes.objectOf(
      PropTypes.shape({
        id: PropTypes.string.isRequired,
        parentId: PropTypes.string,
        children: PropTypes.arrayOf(PropTypes.string).isRequired,
        hasChildren: PropTypes.bool.isRequired,
        hasParent: PropTypes.bool.isRequired,
        isChecked: PropTypes.bool.isRequired,
        isExpanded: PropTypes.bool.isRequired,
        isChildrenLoading: PropTypes.bool.isRequired,
        type: PropTypes.string.isRequired,
        data: PropTypes.shape({
          title: PropTypes.string.isRequired,
        }).isRequired,
      }),
    ).isRequired,
  }),
  padding: PropTypes.number,

  /**
   * Callback called on each modification of an item.
   *
   * @param {object} item The item modified.
   * @param {object} tree The current tree.
   */
  onItemChange: PropTypes.func,
};

const defaultProps = {
  tree: null,
  padding: 16,
  onItemChange: () => {},
};

class LayerTree extends PureComponent {
  static areOthersSiblingsUncheck(tree, item) {
    const parent = tree.items[item.parentId];
    if (!parent) {
      return false;
    }
    return !parent.children
      .filter(id => id !== item.id)
      .find(id => tree.items[id].isChecked);
  }

  static areAllSiblingsUncheck(tree, item) {
    const parent = tree.items[item.parentId];
    if (!parent) {
      return false;
    }
    return !parent.children.find(id => tree.items[id].isChecked);
  }

  constructor(props) {
    super(props);
    this.state = {
      tree: { ...props.tree },
    };
  }

  onExpand(item) {
    const { tree } = this.state;
    let newTree = tree;

    if (item.type === 'radio') {
      newTree = this.mutateTree(newTree, item.id, {
        isExpanded: true,
        isChecked: true,
      });

      if (item.parentId) {
        newTree = this.mutateTree(newTree, item.parentId, {
          isChecked: true,
        });
        tree.items[item.parentId].children.forEach(childId => {
          const child = tree.items[childId];

          if (child.id === item.id) {
            return;
          }

          const val = false;
          if (child.isChecked === val || child.isExpanded === val) {
            return;
          }

          newTree = this.mutateTree(newTree, childId, {
            isExpanded: val,
            isChecked: val,
          });
        });
      }
    } else {
      newTree = this.mutateTree(newTree, item.id, {
        isExpanded: true,
      });
    }

    this.setState({
      tree: newTree,
    });
  }

  onCollapse(item) {
    const { tree } = this.state;
    this.setState({
      tree: this.mutateTree(tree, item.id, { isExpanded: false }),
    });
  }

  onChecked(item, value) {
    const { tree } = this.state;
    let newTree = tree;

    // If input radio change the state of others linked radio input
    // An input radio can only be checked by a user click.
    if (item.type === 'radio' && value) {
      newTree = this.mutateTree(newTree, item.id, {
        isChecked: value,
        isExpanded: item.hasChildren,
      });

      // Check  parents if it is not checked.
      newTree = this.applyToParents(newTree, item, {
        isChecked: value,
      });

      // Uncheck all the radio inputs of the same group.
      newTree = this.applyToChildren(
        newTree,
        newTree.items[item.parentId],
        {
          isChecked: !value,
          isExpanded: !value,
        },
        item,
      );
    } else if (item.type === 'checkbox') {
      const mutation = { isChecked: value };
      newTree = this.mutateTree(newTree, item.id, mutation);

      // Apply to parents if all the others siblings are uncheck.
      newTree = this.applyToParents(newTree, item, mutation);

      // On check, check all the childrens
      // On uncheck, uncheck all the childrens
      newTree = this.applyToChildren(newTree, item, mutation);
    }

    this.setState({
      tree: newTree,
    });
  }

  applyToItem(tree, item, mutation) {
    let newTree = tree;
    const newMutation = { ...mutation };

    // We remove all the unecessary mutations.
    Object.keys({ ...mutation }).forEach(key => {
      if (item[key] === mutation[key]) {
        delete newMutation[key];
      }
    });

    // No mutation to apply
    if (!Object.keys(newMutation).length) {
      return newTree;
    }

    newTree = this.mutateTree(newTree, item.id, newMutation);
    return newTree;
  }

  /**
   * Apply a mutation to all the parents recursively.
   */
  applyToParents(tree, item, mutation) {
    let newTree = tree;
    const parent = newTree.items[item.parentId];

    if (!parent) {
      return newTree;
    }

    // Apply to parents if all the others siblings are uncheck.
    if (LayerTree.areOthersSiblingsUncheck(newTree, item)) {
      newTree = this.applyToItem(newTree, parent, mutation);
      newTree = this.applyToParents(newTree, parent, mutation);
    }
    return newTree;
  }

  /**
   * Apply a mutation to all the chidlren recursively.
   */
  applyToChildren(tree, item, mutation, itemIgnored) {
    let newTree = tree;

    // Go through all the children.
    tree.items[item.id].children.forEach(childId => {
      const child = newTree.items[childId];
      if (itemIgnored && child.id === itemIgnored.id) {
        return;
      }
      newTree = this.applyToItem(newTree, child, mutation);

      if (child.hasChildren) {
        newTree = this.applyToChildren(newTree, child, mutation, itemIgnored);
      }
    });
    return newTree;
  }

  mutateTree(tree, itemId, mutation) {
    const { onItemChange } = this.props;
    const newTree = mutateTree(tree, itemId, mutation);
    onItemChange(newTree.items[itemId], newTree);
    return newTree;
  }

  renderItem({ item, onExpand, onCollapse, provided }) {
    return (
      <div
        ref={provided.innerRef}
        {...provided.draggableProps}
        {...provided.dragHandleProps}
      >
        <input
          type={item.type}
          name={item.parentId}
          checked={item.isChecked}
          onChange={() => {
            // this.onChecked(item, evt.target.checked);
          }}
          onClick={evt => {
            this.onChecked(item, evt.target.checked);
          }}
        />
        <button
          type="button"
          onClick={() => {
            if (!item.hasChildren) {
              return;
            }
            item.isExpanded ? onCollapse(item) : onExpand(item);
          }}
        >
          {item.data.title}
        </button>
      </div>
    );
  }

  render() {
    const { tree } = this.state;
    const { padding } = this.props;
    return (
      <Tree
        tree={tree}
        renderItem={(item, onExpand, onCollapse, provided) =>
          this.renderItem(item, onExpand, onCollapse, provided)
        }
        onExpand={item => this.onExpand(item)}
        onCollapse={item => this.onCollapse(item)}
        offsetPerLevel={padding}
      />
    );
  }
}

LayerTree.propTypes = propTypes;
LayerTree.defaultProps = defaultProps;

export default LayerTree;
