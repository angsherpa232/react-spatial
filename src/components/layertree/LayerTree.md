#

This demonstrates the use of LayerTree.

```jsx
const React = require('react');
const LayerTree = require('./LayerTree').default;
const BasicMap = require('../map/BasicMap').default;
const Layer = require('../../Layer').default;
const VectorLayer = require('../../VectorLayer').default;
const OLMap = require('ol/Map').default;
const Feature = require('ol/Feature').default;
const Point = require('ol/geom/Point').default;
const VectorSource = require('ol/source/Vector').default;
const TileLayer = require('ol/layer/Tile').default;
const TileGrid = require('ol/tilegrid/TileGrid').default;
const TileImageSource = require('ol/source/TileImage').default;
const getCenter = require('ol/extent').getCenter;
const Style = require('ol/style/Style').default;
const Circle = require('ol/style/Circle').default;
const Fill = require('ol/style/Fill').default;
const Text = require('ol/style/Text').default;
require('./LayerTree.md.scss');

class LayerTreeExample extends React.Component {
  constructor(props) {
    super(props);

    const extent = [599500, 199309, 600714, 200002];
    const resolutions = [
      6.927661,
      3.4638305,
      1.73191525,
      0.865957625,
      0.4329788125,
      0.21648940625,
      0.108244703125,
    ];

    const layer = new Layer({
      name: 'Layer',
      olLayer: new TileLayer({
        extent: extent,
        source: new TileImageSource({
          tileUrlFunction: c =>
            '//plans.trafimage.ch/static/tiles/' +
            `bern_aussenplan/${c[0]}/${c[1]}/${c[2]}.png`,
          tileGrid: new TileGrid({
            origin: [extent[0], extent[1]],
            resolutions: resolutions,
          }),
        }),
      }),
    });
    const vectorLayer = new VectorLayer({
      source: new VectorSource({
        features: [
          new Feature({
            geometry: new Point(getCenter(extent)),
          }),
        ],
      }),
      style: new Style({
        image: new Circle({
          radius: 40,
          fill: new Fill({
            color: '#ff0000',
          }),
        }),
      }),
    });
    this.center = getCenter(extent);
    this.layers = [layer, vectorLayer];
    this.map = new OLMap();

    this.state = {
      treeData: { ...treeData },
    };
  }

  applyStyle(item) {
    this.layers[1].olLayer.getStyle().setText(
      new Text({
        font: '20px sans-serif',
        text: 'Last item modified is : ' + item.data.title,
      }),
    );
    this.layers[1].olLayer.changed();
  }

  render() {
    const { featureClicked, treeData } = this.state;

    return (
      <div className="tm-container">
        <BasicMap
          map={this.map}
          center={this.center}
          zoom={17}
          layers={this.layers}
        />
        <LayerTree
          tree={treeData}
          onUpdate={newTreeData => {
            this.setState({
              treeData: newTreeData,
            });
          }}
          onItemChange={item => {
            this.applyStyle(item);
          }}
        />
      </div>
    );
  }
}

<LayerTreeExample />;
```
