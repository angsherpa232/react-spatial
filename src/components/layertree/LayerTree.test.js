import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import LayerTree from './LayerTree';
import data, { applyDefaultValues } from '../../../data/TreeData.esm';

configure({ adapter: new Adapter() });

const inputSelector = 'input';
const expandButtonSelector = 'button';

describe('LayerTree', () => {
  let mutations;

  beforeEach(() => {
    mutations = [];
  });

  const mountLayerTree = data2 =>
    mount(
      <LayerTree
        tree={applyDefaultValues(data2)}
        onItemChange={(item, tree) => {
          mutations.push({ item, tree });
        }}
      />,
    );

  describe('matches snapshots', () => {
    test('using default properties.', () => {
      // Test-renderer cannot be use because of th DnD functionnality which needs an existing Html element
      const component = mount(<LayerTree tree={data} />);
      expect(component.html()).toMatchSnapshot();
    });

    test('when renderItem is used.', () => {
      // Test-renderer cannot be use because of th DnD functionnality which needs an existing Html element
      const component = mount(
        <LayerTree tree={data} renderItem={item => <div>{item.title}</div>} />,
      );
      expect(component.html()).toMatchSnapshot();
    });

    test('when an item is hidden.', () => {
      // Test-renderer cannot be use because of th DnD functionnality which needs an existing Html element
      const component = mount(
        <LayerTree
          tree={data}
          isItemHidden={item => {
            return !!item.children.length;
          }}
        />,
      );
      expect(component.html()).toMatchSnapshot();
    });
  });

  describe('when an checkbox item is checked', () => {
    test('check all children', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1', '2'],
          },
          '1': {
            children: ['1-1', '1-2'],
          },
          '2': {},
          '1-1': {},
          '1-2': {
            children: ['1-2-1'],
          },
          '1-2-1': {},
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(inputSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(5);
      const lastItems = mutations.pop().tree.items;
      ['1', '1-1', '1-2', '1-2-1'].forEach(id => {
        expect(lastItems[id].isChecked).toBe(true);
      });
      ['2'].forEach(id => {
        expect(lastItems[id].isChecked).toBe(false);
      });
    });

    describe('when siblings are uncheck', () => {
      test('check all parents', () => {
        const data2 = {
          rootId: 'root',
          items: {
            root: {
              children: ['1', '2'],
            },
            '1': {
              isExpanded: true,
              children: ['1-1', '1-2'],
            },
            '2': {},
            '1-1': {},
            '1-2': {
              isExpanded: true,
              children: ['1-2-1', '1-2-2'],
            },
            '1-2-1': {},
            '1-2-2': {},
          },
        };
        const component = mountLayerTree(data2);
        component
          .find(inputSelector)
          .at(4) // 1-2-1
          .simulate('click');
        expect(mutations.length).toBe(4);
        const lastItems = mutations.pop().tree.items;
        ['root', '1', '1-2', '1-2-2'].forEach(id => {
          expect(lastItems[id].isChecked).toBe(true);
        });
        ['2', '1-1', '1-2-1'].forEach(id => {
          expect(lastItems[id].isChecked).toBe(false);
        });
      });
    });
  });

  describe('when a radio item is checked', () => {
    test('expands the item, uncheck and collapse all the siblings, and apply defaults value to the children.', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1', '2', '3'],
          },
          '1': {
            type: 'radio',
            children: ['1-1', '1-2'],
          },
          '2': {},
          '3': {
            isChecked: true,
            isExpanded: true,
            type: 'radio',
          },
          '1-1': {
            defaults: {
              isChecked: true,
              isExpanded: true,
            },
          },
          '1-2': {
            children: ['1-2-1'],
          },
          '1-2-1': {},
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(inputSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(3);
      const lastItems = mutations.pop().tree.items;
      ['1', '1-1'].forEach(id => {
        expect(lastItems[id].isChecked).toBe(true);
        expect(lastItems[id].isExpanded).toBe(true);
      });
      ['2', '3', '1-2', '1-2-1'].forEach(id => {
        expect(lastItems[id].isChecked).toBe(false);
        expect(lastItems[id].isExpanded).toBe(false);
      });
    });
  });

  describe('when an radio item is unchecked', () => {
    test('uncheck all children', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1', '2', '3'],
          },
          '1': {
            isChecked: true,
            isExpanded: true,
            type: 'radio',
            children: ['1-1', '1-2', '1-3'],
          },
          '1-1': {
            type: 'radio',
          },
          '1-2': {
            isExpanded: true,
            children: ['1-2-1'],
          },
          '1-2-1': {
            isChecked: true,
          },
          '1-3': {
            isChecked: true,
            isExpanded: true,
            type: 'radio',
          },
          '2': {},
          '3': {
            isChecked: false,
            isExpanded: false,
            type: 'radio',
          },
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(inputSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(3);
      const lastItems = mutations.pop().tree.items;
      ['1', '1-1', '1-2', '1-3', '1-2-1', '2', '3'].forEach(id => {
        expect(lastItems[id].isChecked).toBe(false);
      });
    });

    describe('when siblings are uncheck', () => {
      test('uncheck all parents.', () => {
        const data2 = {
          rootId: 'root',
          items: {
            root: {
              type: 'radio',
              children: ['1', '2'],
            },
            '1': {
              type: 'radio',
              isChecked: true,
              isExpanded: true,
              children: ['1-1', '1-2'],
            },
            '1-1': {
              type: 'radio',
            },
            '1-2': {
              type: 'radio',
              isChecked: true,
              isExpanded: true,
              children: ['1-2-1', '1-2-2'],
            },
            '1-2-1': {},
            '1-2-2': {},
            '2': {},
          },
        };
        const component = mountLayerTree(data2);
        component
          .find(inputSelector)
          .at(2) // 1-2
          .simulate('click');
        expect(mutations.length).toBe(2);
        const lastItems = mutations.pop().tree.items;
        ['1', '1-2'].forEach(id => {
          expect(lastItems[id].isChecked).toBe(false);
        });
      });
    });
  });

  describe('when an checkbox item is unchecked', () => {
    test('uncheck all children', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            isChecked: true,
            children: ['1', '2', '3'],
          },
          '1': {
            isChecked: true,
            isExpanded: true,
            children: ['1-1', '1-2', '1-3'],
          },
          '1-1': {
            type: 'radio',
          },
          '1-2': {
            isExpanded: true,
            children: ['1-2-1'],
          },
          '1-2-1': {
            isChecked: true,
          },
          '1-3': {
            isChecked: true,
            isExpanded: true,
            type: 'radio',
          },
          '2': {},
          '3': {
            isChecked: false,
            isExpanded: false,
            type: 'radio',
          },
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(inputSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(4);
      const lastItems = mutations.pop().tree.items;
      ['root', '1', '1-2-1', '1-3'].forEach(id => {
        expect(lastItems[id].isChecked).toBe(false);
      });
    });

    describe('when siblings are uncheck', () => {
      test('uncheck all parents.', () => {
        const data2 = {
          rootId: 'root',
          items: {
            root: {
              children: ['1', '2'],
            },
            '1': {
              type: 'radio',
              isChecked: true,
              isExpanded: true,
              children: ['1-1', '1-2'],
            },
            '1-1': {
              type: 'radio',
            },
            '1-2': {
              isChecked: true,
              isExpanded: true,
              children: ['1-2-1', '1-2-2'],
            },
            '1-2-1': {},
            '1-2-2': {},
            '2': {},
          },
        };
        const component = mountLayerTree(data2);
        component
          .find(inputSelector)
          .at(2) // 1-2
          .simulate('click');
        expect(mutations.length).toBe(2);
        const lastItems = mutations.pop().tree.items;
        ['1', '1-2'].forEach(id => {
          expect(lastItems[id].isChecked).toBe(false);
        });
      });
    });
  });

  describe('expands', () => {
    test('a radio item', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1'],
          },
          '1': {
            type: 'radio',
            children: ['1-1'],
          },
          '1-1': {},
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(expandButtonSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(1);
      const lastItems = mutations.pop().tree.items;
      expect(lastItems['1'].isExpanded).toBe(true);
      expect(lastItems['1'].isChecked).toBe(false);
    });

    test('a checkboy item', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1'],
          },
          '1': {
            children: ['1-1'],
          },
          '1-1': {},
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(expandButtonSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(1);
      const lastItems = mutations.pop().tree.items;
      expect(lastItems['1'].isExpanded).toBe(true);
      expect(lastItems['1'].isChecked).toBe(false);
    });
  });

  describe('collapse', () => {
    test('a radio item', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1'],
          },
          '1': {
            type: 'radio',
            isExpanded: true,
            isChecked: true,
            children: ['1-1'],
          },
          '1-1': {
            isChecked: true,
          },
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(expandButtonSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(1);
      const lastItems = mutations.pop().tree.items;
      expect(lastItems['1'].isExpanded).toBe(false);
      expect(lastItems['1'].isChecked).toBe(true);
    });

    test('a checkboy item', () => {
      const data2 = {
        rootId: 'root',
        items: {
          root: {
            children: ['1'],
          },
          '1': {
            isExpanded: true,
            isChecked: true,
            children: ['1-1'],
          },
          '1-1': {
            isChecked: true,
          },
        },
      };
      const component = mountLayerTree(data2);
      component
        .find(expandButtonSelector)
        .first()
        .simulate('click');
      expect(mutations.length).toBe(1);
      const lastItems = mutations.pop().tree.items;
      expect(lastItems['1'].isExpanded).toBe(false);
      expect(lastItems['1'].isChecked).toBe(true);
      expect(lastItems['1-1'].isChecked).toBe(true);
    });
  });
});
