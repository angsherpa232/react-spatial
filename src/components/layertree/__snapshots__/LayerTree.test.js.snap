// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LayerTree matches snapshots using default properties. 1`] = `"<div style=\\"pointer-events: auto;\\" data-react-beautiful-dnd-droppable=\\"0\\"><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"0\\" style=\\"padding-left: 0px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generaterootradio\\"><span></span></label><div>ovnetzkarteschweiz</div><button class=\\"tm-arrow tm-arrow-up\\" type=\\"button\\"></button></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"0\\" style=\\"padding-left: 30px;\\"><label class=\\"tm-layertree-input-checkbox\\" tabindex=\\"0\\"><input type=\\"checkbox\\" name=\\"shortid.generateovnetzkarteschweizcheckbox\\"><span></span></label><div>bahnhofplane</div><button class=\\"tm-arrow tm-arrow-up\\" type=\\"button\\"></button></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"0\\" style=\\"padding-left: 60px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generatebahnhofplaneradio\\"><span></span></label><div>interaktivbahnhoplan</div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"0\\" style=\\"padding-left: 60px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generatebahnhofplaneradio\\"><span></span></label><div>printprodukte</div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"0\\" style=\\"padding-left: 30px;\\"><label class=\\"tm-layertree-input-checkbox\\" tabindex=\\"0\\"><input type=\\"checkbox\\" name=\\"shortid.generateovnetzkarteschweizcheckbox\\"><span></span></label><div>passagierfrequenzen</div><button class=\\"tm-arrow tm-arrow-down\\" type=\\"button\\"></button></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"0\\" style=\\"padding-left: 0px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generaterootradio\\"><span></span></label><div>bauprojekte</div></div></div>"`;

exports[`LayerTree matches snapshots when an item is hidden. 1`] = `"<div style=\\"pointer-events: auto;\\" data-react-beautiful-dnd-droppable=\\"2\\"><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"2\\" style=\\"padding-left: 0px;\\"></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"2\\" style=\\"padding-left: 30px;\\"></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"2\\" style=\\"padding-left: 60px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generatebahnhofplaneradio\\"><span></span></label><div>interaktivbahnhoplan</div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"2\\" style=\\"padding-left: 60px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generatebahnhofplaneradio\\"><span></span></label><div>printprodukte</div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"2\\" style=\\"padding-left: 30px;\\"></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"2\\" style=\\"padding-left: 0px;\\"><label class=\\"tm-layertree-input-radio\\" tabindex=\\"0\\"><input type=\\"radio\\" name=\\"shortid.generaterootradio\\"><span></span></label><div>bauprojekte</div></div></div>"`;

exports[`LayerTree matches snapshots when renderItem is used. 1`] = `"<div style=\\"pointer-events: auto;\\" data-react-beautiful-dnd-droppable=\\"1\\"><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"1\\" style=\\"padding-left: 0px;\\"><div></div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"1\\" style=\\"padding-left: 30px;\\"><div></div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"1\\" style=\\"padding-left: 60px;\\"><div></div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"1\\" style=\\"padding-left: 60px;\\"><div></div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"1\\" style=\\"padding-left: 30px;\\"><div></div></div><div class=\\"tm-layertree-item\\" data-react-beautiful-dnd-draggable=\\"1\\" style=\\"padding-left: 0px;\\"><div></div></div></div>"`;
